src/Common/ThreadPool.h

```diff
@@ -252,6 +265,9 @@ class ThreadFromGlobalPoolImpl : boost::noncopyable
 
         /// The state used in this object and inside the thread job.
         Poco::Event event;
+
+        /// To allow joining to the same
+        std::atomic<bool> finished{true};
```

Comment:
AFAIU, we should initialized `finished` with `false` there, otherwise the flag is redundant and we can always true from `joinable` if `state` is reseted.

Reply:
Nice catch. It was previously named `is_running` and I forgot to update initial value...